### PR TITLE
[DI2-146] Handle legacy drawing parameters for drawings

### PR DESCRIPTION
**Pull request by jonkoops**

### CODE REVIEW

File: src/app/pages/MapPage/query-params.ts [hunk 1]
```diff
@@ -166,6 +139,73 @@ export const panoTagParam: UrlParam<string> = {
...
+  return {
+    id: parsedValue.id,
+    polygon: parsedValue.polygon.map(([lat, lng]) => ({ lat, lng })),
+  }
+}
+
+function encodePolyDrawing(value: PolyDrawing) {
+  const encodedValue: RawPolyDrawing = {
+    id: value.id,
+    polygon: value.polygon.map(({ lat, lng }) => [lat, lng]),
```
Comment (janjaap) [inline src/app/pages/MapPage/query-params.ts:187]: Why transform an object to an array when you're doing the opposite in the decoding function? Wouldn't it be easier to leave it as an object?

File: src/app/pages/MapPage/query-params.test.ts [hunk 2]
```diff
@@ -1,29 +1,65 @@
...
+    const description = '1,21 km en 59.151,1 m²'
+    const legacyValue = `{"description":"${description}","markers":"52.37628201724842:4.896165591846303|52.37422313732434:4.891254241303624|52.37422452094195:4.898852807226181"}`
+    const expectedLegacy = {
+      id: btoa(description),
+      polygon: expectedDecoded.polygon,
+    }
+
+    expect(polygonParam.decode(legacyValue)).toEqual(expectedLegacy)
+    expect(polylineParam.decode(legacyValue)).toEqual(expectedLegacy)
   })
```
Comment (janjaap) [inline src/app/pages/MapPage/query-params.test.ts:64]: Well done on the coverage

File: src/app/pages/MapPage/draw/DrawTool.tsx [hunk 3]
```diff
@@ -20,12 +20,15 @@ import {
 import { useHistory } from 'react-router-dom'
```
Comment (janjaap) [inline src/app/pages/MapPage/draw/DrawTool.tsx:20]: The `DrawTool` component doesn't have any tests. Can you add tests for the parts of this component that you changed?

File: src/app/pages/MapPage/MapContainer.tsx [hunk 4]
```diff
@@ -47,9 +47,9 @@ const reducer = (state: MapState, action: Action): MapState => {
 const MapContainer: FunctionComponent = ({ children }) => {
   const [state, dispatch] = useReducer(reducer, initialState)
   const [activeMapLayers] = useParam(mapLayersParam)
-  const [polylines] = useParam(polylinesParam)
-  const [polygons] = useParam(polygonsParam)
-  const showDrawContent = polygons.length > 0 || polylines.length > 0
+  const [polyline] = useParam(polylineParam)
+  const [polygon] = useParam(polygonParam)
+  const showDrawContent = !!polyline && !!polygon
```
Comment (janjaap) [inline src/app/pages/MapPage/MapContainer.tsx:52]: As far as I can see, the value of either `polylineParam` or `polylineParam` is a string or `null`. There shouldn't be any need for a double negation operator.

File: src/app/pages/MapPage/query-params.ts (hunk 1 above)
Comment (jonkoops) [reply src/app/pages/MapPage/query-params.ts:187]: The reason is to reduce the amount of text that is used inside of the URL. Although the URL length limit is no longer a problem in most modern browsers it's still best to keep things as compact as possible.

File: src/app/pages/MapPage/query-params.test.ts (hunk 2 above)
Comment (jonkoops) [reply src/app/pages/MapPage/query-params.test.ts:64]: Thanks :D

File: src/app/pages/MapPage/MapContainer.tsx (hunk 4 above)
Comment (jonkoops) [reply src/app/pages/MapPage/MapContainer.tsx:52]: I thought the same, however TypeScript infers the type of `showDrawContent` as `PolyDrawing | null` if I do not do this casting explicitly.

image

File: src/app/pages/MapPage/draw/DrawTool.tsx (hunk 3 above)
Comment (jonkoops) [reply src/app/pages/MapPage/draw/DrawTool.tsx:20]: Yeah sure, I will take a look if I can do something tomorrow.

Comment (jonkoops) [reply src/app/pages/MapPage/draw/DrawTool.tsx:20]: Due to the many dependencies this one is a bit of a hassle to cover right now. I've created a ticket on our backlog to improve the coverage of drawtool related code.

File: src/app/pages/MapPage/draw/DrawResults.tsx [hunk 5]
```diff
@@ -294,80 +285,55 @@ const DrawResults: FunctionComponent<Props> = ({ currentOverlay }) => {
...
                     />
                   )}
-                </AccordionContent>
-                {totalCount === 0 && (
-                  <StyledAlert level="info">Er zijn geen resultaten</StyledAlert>
-                )}
-              </StyledAccordion>
+                </>
+              )}
+              {delayedLoadingIds.includes(id) && !errorIds.includes(id) && (
```
Comment (chrisvanmook) [inline src/app/pages/MapPage/draw/DrawResults.tsx:324]: perhaps we could move these statements to a variable so it's more readable?

File: src/app/pages/MapPage/query-params.test.ts [hunk 6]
```diff
@@ -1,29 +1,65 @@
...
+  it('decodes the parameter', () => {
+    expect(polygonParam.decode(expectedEncoded)).toEqual(expectedDecoded)
+    expect(polylineParam.decode(expectedEncoded)).toEqual(expectedDecoded)
+  })
+
+  it('decodes the parameter with a legacy value', () => {
+    const description = '1,21 km en 59.151,1 m²'
+    const legacyValue = `{"description":"${description}","markers":"52.37628201724842:4.896165591846303|52.37422313732434:4.891254241303624|52.37422452094195:4.898852807226181"}`
+    const expectedLegacy = {
+      id: btoa(description),
```
Comment (chrisvanmook) [inline src/app/pages/MapPage/query-params.test.ts:58]: `description` is used for an `id`? 🤔

Comment (jonkoops) [reply src/app/pages/MapPage/query-params.test.ts:58]: Yeah, these need to be consistent to keep track of the layers, for more info see [this comment ](https://github.com/Amsterdam/atlas/pull/3931/files#diff-6993cba0fc3a3d48d5bd3ffb4f410f7111b91193488f412a799f48e70c2e464bR194-R195)in the code.

File: src/app/pages/MapPage/draw/DrawResults.tsx [hunk 7]
```diff
@@ -113,11 +104,11 @@ const StyledSelect = styled(Select)`
...
 `
 
-type Props = {
+export interface DrawResultsProps {
   currentOverlay: Overlay
 }
 
-const DrawResults: FunctionComponent<Props> = ({ currentOverlay }) => {
+const DrawResults: FunctionComponent<DrawResultsProps> = ({ currentOverlay }) => {
   const [delayedLoadingIds, setDelayedLoadingIds] = useState<string[]>([])
```
[the remark this answers is not in the document]
Comment (jonkoops) [reply src/app/pages/MapPage/draw/DrawResults.tsx:112]: This code was there originally already, so I'd like to keep it like this.

File: src/app/pages/MapPage/draw/DrawResults.tsx (hunk 5 above)
Comment (jonkoops) [reply src/app/pages/MapPage/draw/DrawResults.tsx:324]: This code was there originally already, so I'd like to keep it like this.

**Review by chrisvanmook — APPROVED** [review]

**Pull request merged**